### ISSUE TITLE
Solaris 11 fails for aes-test

### ISSUE THREAD

**Issue by quanah** [heimdal/heimdal#314]
make check fails on Solaris 11:

```
make  check-TESTS check-local
make[4]: Entering directory `/export/home/quanah/git/sold-2445/SunOS5.11-i86pc/build/heimdal.i86pc_64/lib/krb5'
aes-test: 3: key wrong
aes-test: 5: key wrong
aes-test: 7: key wrong
aes-test: 9: key wrong
aes-test: 11: key wrong
aes-test: 13: key wrong
aes-test: 15: key wrong
aes-test: krb5_decrypt: Decrypt integrity check failed for checksum type hmac-sha1-96-aes256, key type aes256-cts-hmac-sha1-96
aes-test: krb_enc failed with -1765328353 for test 0
FE697B52BC0D3CE14432BA036A92E65BF6F26A682C3624E7EDC7E03F09D722E3
FE697B52BC0D3CE14432BA036A92E65BBB52280990A2FA27883998D72AF30161
A2E16D16B36069C135D5E9D2E25F896158F09858BBBEA122F01184B8F4F76566
A2E16D16B36069C135D5E9D2E25F896102685618B95914B467C67622225824FF
55A6AC740AD17B4846941051E1E8B0A7B7319199D6D3FFFF82EF3FD0AFCF6F85
55A6AC740AD17B4846941051E1E8B0A7548D93B0AB30A8BC3FF16280382B8C2A
97A4E786BE20D81A382D5EBC96D5909C87FA858C57532E9D16AA08321347716F
97A4E786BE20D81A382D5EBC96D5909CABCDADC87CA48F574504159F16C36E31
89ADEE3608DB8BC71F1BFBFE459486B0A02F28E4F1E927EC7B197C63D12E8251
89ADEE3608DB8BC71F1BFBFE459486B05618B70CBAE22092534E56C553BA4B34
D78C5C9CB872A8C9DAD4697F0BB5B2D2189D53C487974F88E22D8654C2AEA0C5
D78C5C9CB872A8C9DAD4697F0BB5B2D21496C82BEB2CAEDA2112FCEEA057401B
4B6D9839F84406DF1F09CC166DB4B83CAE95E4013CAE6E6B7FF905A65A1DFB23
4B6D9839F84406DF1F09CC166DB4B83C571848B784A3D6BDC346589A3E393F9E
FAIL: aes-test
```

**Comment by sfhacker** [comment]
Just built master branch from source and got this:

PASS aes-test.exe (exit status: 0)

P.S.: Try to provide more info so it is easier to reproduce this isse!

**Comment by quanah** [comment]
7.4.0 release, x86-64 architecture, using the GNU build tools.

**Comment by quanah** [comment]
@sfhacker You clearly built on Windows (aes-test.exe) and not on Solaris 11.  As I noted in my report, this issue is *specific* to Solaris 11.

**Comment by quanah** [comment]
@nicowilliams What additional data can I collect to help move this along?

**Comment by nicowilliams** [comment]
@quanah Is this using the PKCS#11 provider?

**Comment by quanah** [comment]
@nicowilliams I'm building out Solaris 11 the exact same way I do on Linux, building and linking Heimdal to OpenSSL 1.0.2k for the crypto libraries.

**Comment by SethEllsworth** [comment]
I get the same 4 make check failures (aes-test,test_fx,string-to-key-test, derived-key-test ) in lib/krb5 on both Solaris 10 Sparc and Solaris 10 x64.

Was any progress make on fixing this?
We are currently using the heimdal-7-1-branch at the 7.4 release.

Seems they are all related to AES testing.
-bash-3.2$ ./test_fx
test_fx: key not same for enctypes 18 18 18
test_fx: key not same for enctypes 18 18 17
test_fx: key not same for enctypes 17 18 18
test_fx: KRB-FX-CF2 not supported for enctype 1: kerberos prf for des-cbc-crc not supported
test_fx: key not same for enctypes 16 16 16
test_fx: 4 KRB-FX-CF2 vectors failed
-bash-3.2$ grep "AES.* = 1[678]" ../asn1/krb5_asn1.h
  CKSUMTYPE_HMAC_SHA1_96_AES_256 = 16,
  KRB5_ENCTYPE_AES128_CTS_HMAC_SHA1_96 = 17,
  KRB5_ENCTYPE_AES256_CTS_HMAC_SHA1_96 = 18,
-bash-3.2$

Actual string output:
-bash-3.2$ ./string-to-key-test
...
string_to_key(nisse@FOO.SE, hej) failed
should be: 0ebc239d6846f2d551985b57c157017904c4e9fec10e13d0
result was: 0ebc239d6846f2d55164a4a42c2acdbcec1c64a4a42c2a7f
...

Odd how the first 18 characters match, then it differs after that. Maybe something with block processing? 

Any hints/help/places to look next are welcome, I've never dived into the actual crypto processing before.

**Comment by SethEllsworth** [comment]
It seems related to salting. I was looking at the wrong spot for enctypes:
  KRB5_ENCTYPE_DES3_CBC_SHA1 = 16,
  KRB5_ENCTYPE_AES128_CTS_HMAC_SHA1_96 = 17,
  KRB5_ENCTYPE_AES256_CTS_HMAC_SHA1_96 = 18,

All salted. 

ARCFOUR doesn't seem to be failing, and it isn't salted.

**Comment by lhoward** [comment]
Reproduced, investigating.

**Comment by lhoward** [comment]
Appears related to the fix introduced in #194.

**Comment by lhoward** [comment]
#194 shouldn't have broken things, but it's appearing that second and subsequent calls to C_EncryptUpdate() are NOOPs (at least on my Solaris 11 system). Attempting to determine cause.

Calling C_Initialize()/C_Encrypt() for each invocation of p11_do_cipher() works, although it does break the case noted in #194 (the problem there was IV chaining, which I have a workaround for, but because the workaround is an abstraction violation I'm trying to fix it properly).

**Comment by lhoward** [comment]
Needs a better commit message, but see 2b5b7cf for now.

**Comment by quanah** [comment]
With 2b5b7cf applied, aes-test passes for me.

**Comment by lhoward** [comment]
Fixed in a88d0066.
